cmd/tempo-vulture/main.go

```diff
@@ -45,95 +42,148 @@ func init() {
 
-	flag.StringVar(&lokiBaseURL, "loki-base-url", "", "The base URL (scheme://hostname) at which to find loki.")
-	flag.StringVar(&lokiQuery, "loki-query", "", "The query to use to find traceIDs in Loki.")
-	flag.StringVar(&lokiUser, "loki-user", "", "The user to use for Loki basic auth.")
-	flag.StringVar(&lokiPass, "loki-pass", "", "The password to use for Loki basic auth.")
-
-	flag.StringVar(&tempoBaseURL, "tempo-base-url", "", "The base URL (scheme://hostname) at which to find tempo.")
+	flag.StringVar(&tempoQueryURL, "tempo-query-url", "", "The URL (scheme://hostname) at which to query Tempo.")
+	flag.StringVar(&tempoPushURL, "tempo-push-url", "", "The URL (scheme://hostname) at which to push traces to Tempo.")
 	flag.StringVar(&tempoOrgID, "tempo-org-id", "", "The orgID to query in Tempo")
-	flag.DurationVar(&tempoBackoffDuration, "tempo-backoff-duration", time.Second, "The amount of time to pause between tempo calls")
+	flag.DurationVar(&tempoBackoffDuration, "tempo-backoff-duration", time.Second, "The amount of time to pause between Tempo calls")
 }
 
 func main() {
 	flag.Parse()
 
-	glog.Error("Application Starting")
+	glog.Error("Tempo Vulture Starting")
 
-	testDurations := []time.Duration{
-		24 * time.Hour,
-		12 * time.Hour,
-		6 * time.Hour,
-		3 * time.Hour,
-		time.Hour,
-		30 * time.Minute,
-	}
+	startTime := time.Now().UTC().UnixNano()
+	ticker := time.NewTicker(tempoBackoffDuration)
+	iteration := int64(0)
 
-	ticker := time.NewTicker(15 * time.Second)
+	// Write
 	go func() {
 		for {
 			<-ticker.C
 
-			for _, duration := range testDurations {
+			iteration++
+			rand.Seed(startTime / iteration)
 
-				// query loki for trace ids
-				lines, err := queryLoki(lokiBaseURL, lokiQuery, duration, lokiUser, lokiPass)
-				if err != nil {
-					glog.Error("error querying Loki ", err)
-					metricErrorTotal.Inc()
-					continue
-				}
-				ids := extractTraceIDs(lines)
-
-				// query tempo for trace ids
-				metrics, err := queryTempoAndAnalyze(tempoBaseURL, tempoBackoffDuration, ids)
-				if err != nil {
-					glog.Error("error querying Tempo ", err)
-					metricErrorTotal.Inc()
-					continue
-				}
+			c, err := newJaegerGRPCClient(tempoPushURL)
+			if err != nil {
+				glog.Error("error creating grpc client", err)
+				metricErrorTotal.Inc()
+				continue
+			}
 
-				metricTracesInspected.WithLabelValues(strconv.Itoa(int(duration.Seconds()))).Add(float64(metrics.requested))
-				metricTracesErrors.WithLabelValues("notfound", strconv.Itoa(int(duration.Seconds()))).Add(float64(metrics.notfound))
-				metricTracesErrors.WithLabelValues("missingspans", strconv.Itoa(int(duration.Seconds()))).Add(float64(metrics.missingSpans))
+			batch := makeThriftBatch()
+			err = c.EmitBatch(context.Background(), batch)
+			if err != nil {
+				glog.Error("error pushing trace to Tempo ", err)
+				metricErrorTotal.Inc()
+				continue
 			}
 		}
 	}()
 
+	// Read
+	go func() {
+		for {
+			<-ticker.C
```

Comment:
on the read path we can now choose any 5 second interval in the past n days, seed a random number generator, generate a trace and query tempo for it.  

"n days" should be a mix of two parameters
- min-time             : we can just hardcode a datetime to never query before.  this will basically be the time we first start the vulture and will prevent it from querying for traces before it started.
- tempo-retention  : since we eventually drop traces we need to pass the retention to tempo-vulture so it won't query traces after we know tempo should have dropped them.

so the query path would be like:
1. randomly choose a time after max(min-time, tempo-retention) and before now
2. do the same floor operation above ((unix / 5) * 5)
3. generate trace
4. query for trace
5. confirm everything matches up.
6. back to 1

Reply:
Thanks a lot for the explanation! I updated the PR. Hopefully, it's more in-line with your proposal. 

As you can see I re-used the `startTime` variable after retention, but I can create an internal go-routine variable if needed (that way startTime is constant).

Also, I used a different time for the ticker and the slot (slot=ticker*2). That's because I had some rounding problems... And preferred to give it some free room.

Reply:
nice.  yup, this is definitely what i was thinking.

> Also, I used a different time for the ticker and the slot (slot=ticker*2)

I think 2 seconds is just really tight for trace creation.  It could be as loose as 30 seconds and it would be fine.  Maybe  split how often to generate a trace and how often to query them back?

Also, don't be concerned if we see some failures at first start b/c it's querying a trace that does not yet exist.  This doesn't have to be perfect.  It just a tool to help find consistency issues if they creep into the codebase.

Reply:
Oh! My bad. I was using 2 seconds for testing. Bumped it to 30s, sounds like a good default.

Maybe splitting the how often we do each thing is a good idea, but I'm not sure if a lot of people would customize that.